pkg/runtime/serializer/codec_factory.go

```diff
@@ -55,16 +62,78 @@ func newCodecFactory(scheme *runtime.Scheme, mf json.MetaFactory) CodecFactory {
 			Serializer:         jsonSerializer,
 			PrettySerializer:   jsonPrettySerializer,
 		},
-		{
-			AcceptContentTypes: []string{"application/yaml"},
-			ContentType:        "application/yaml",
-			FileExtensions:     []string{"yaml"},
-			Serializer:         yamlSerializer,
-		},
 	}
+	yamlSerializer := json.NewYAMLSerializer(mf, scheme, runtime.ObjectTyperToTyper(scheme))
+	serializers = append(serializers, serializerType{
+		AcceptContentTypes: []string{"application/yaml"},
+		ContentType:        "application/yaml",
+		FileExtensions:     []string{"yaml"},
+		Serializer:         yamlSerializer,
+	})
+
+	for _, fn := range serializerExtensions {
+		if serializer, ok := fn(scheme); ok {
+			serializers = append(serializers, serializer)
+		}
+	}
+	return serializers
+}
+
+// CodecFactory provides methods for retrieving codecs and serializers for specific
+// versions and content types.
+type CodecFactory struct {
+	scheme      *runtime.Scheme
+	serializers []serializerType
+	universal   runtime.Decoder
+	accepts     []string
+
+	legacySerializer runtime.Serializer
+}
+
+// NewCodecFactory provides methods for retrieving serializers for the supported wire formats
+// and conversion wrappers to define preferred internal and external versions. In the future,
+// as the internal version is used less, callers may instead use a defaulting serializer and
+// only convert objects which are shared internally (Status, common API machinery).
+// TODO: allow other codecs to be compiled in?
+// TODO: accept a scheme interface
+func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
+	serializers := newSerializersForScheme(scheme, json.DefaultMetaFactory)
+	return newCodecFactory(scheme, serializers)
+}
+
+// NewStreamingCodecFactory returns serializers that support the streaming.Serializer interface.
+func NewStreamingCodecFactory(scheme *runtime.Scheme) CodecFactory {
```

Comment:
Today, application/json for a watch and regular list call are the same
serializer.  application/yaml or protobuf would not be - when we write
a framed proto message or a "---" delimited yaml object the delimiter
must be written after each object (or before, for proto).  So a
streaming serializer and a regular serializer are not quite the same.
Furthermore, the streaming serializer also needs to be a regular
serializer (object -> yaml doc).

That said, part of the open design question was whether the streaming
codec is a different interface, which would make the one factory
return something different.  The constant call is the mime type, but
what happens on encode and decode is different between streaming vs
not.